Decompose VM instantiation costs and add module cache

graydon opened the pull request:
This is a sketch of the first step in the plan for addressing #1292 and eventually #827 and #1313, namely:

  - add a bunch of cost types that decompose the current "worst case" VM instantiation cost type
  - continue to charge the worst case on initial contract upload
  - _store_ the decomposed cost-type inputs in the ledger, since we can observe them after the initial upload parse
  - _use_ those decomposed cost-type inputs when doing runtime instantiation
  - add a module cache
  - populate the module cache with all modules on host startup
  - use cached modules during instantiation

This PR has accompanying changes in XDR and wasmi:

  - https://github.com/stellar/stellar-xdr/pull/177
  - https://github.com/stellar/rs-stellar-xdr/pull/346

Remaining to do:

  - [x] determine what the correct set of decomposed cost types even is
  - [ ] ~add more code to wasmi to enable observing more inputs~
  - [x] add cost-type runners / calibrations for all the decomposed cost-types
  - [x] protocol-gate this new behaviour
  - [x] make the linker-loop do less work to match the tighter cost model (i.e. complete #1292)
  - [x] possibly _duplicate_ the set of cost types added here, so we have a cached and uncached version of each, and implement a solution to #827 
  - [ ] possibly decompose the set sufficiently to model what will happen when we take wasmi 0.32 and enable lazy translation (i.e. complete #1313)

dmkozh commented on soroban-env-host/src/host/data_helper.rs:
```diff
@@ -122,15 +123,25 @@ impl Host {
+            LedgerEntryData::ContractCode(e) => {
+                let code = e.code.metered_clone(self)?;
+                let costs = match &e.ext {
+                    ContractCodeEntryExt::V0 => None,
```
> So old contracts that gets re-uploaded will automatically have such information.

Wasms can not currently be reuploaded and are keyed by the hash of the uploaded Wasm. So if we do nothing, that contracts uploaded during p20 (and maybe even p21, depending on when we roll this out) will never get the extension, unless we do something. I see a few options:

- Do nothing. If most of the contracts are upgradeable, then the owners could upload the new Wasm (maybe with minor no-op changes to change the hash) and get it cheaper. I'm not a big fan of this option because rent for the Wasm upload is pretty expensive and it's not very nice to make the users to either pay it again, or pay more for every invocation.
- Change upload host function to generate the extension even if Wasm is already present on-chain (currently this is a no-op). This seems pretty nice fee-wise, but would require contract maintainers to manually run this operation, which might be not obvious for everyone.
- Lazily generate the extension on Wasm access. This removes the need for anyone to act explicitly, but will introduce unexpected increased fees for the 'unlucky' users that trigger this code path.

Out of these options I would mildly prefer the second one. But maybe there are more options that I'm missing, which is why I'm curious about others' opinions.

jayz22 replied:
Ah I see. In this case I think both 2 and 3 would work. 
3 is nicer from the operations perspective. The extension gets populated lazily on invocation, so any contract that gets used becomes cheap to invoke. Although not sure if there is an incentive problem as you mentioned. Maybe the extra cpu isn't as bad in terms of fees. 
2 is good from incentives perspective, shifting the cost from the contract user to the contract code owner/maintainer (who has the incentive to make sure their contracts are cheap to invoke).

dmkozh replied:
Actually 3 is kind of risky, at it requires adding a potentially large ledger entry to the RW footprint. Besides the fact that the cost would go up significantly, the transaction might simply exceed the network limit on the write bytes. In the worst case the contract will become completely unusable until it has been updated. So probably we can/should only do 2, even thought it doesn't guarantee full migration.

sisuresh replied:
>  Besides the fact that the cost would go up significantly, the transaction might simply exceed the network limit on the write bytes.

Won't this be an issue either way?

dmkozh replied:
No, if we only keep 'on-demand' extension creation, then it will be just a single operation that consumes roughly the same resources as a regular Wasm upload. But in the 'lazy' case we would add a non-trivial write that is not generally expected by the dapp.

graydon replied:
Implemented the "reupload to update it" change.

graydon commented:
Updated significantly:

  - Does intra-tx caching
  - Does minimal linking
  - Splits parsing from instantiation to, I think, be _more_ ready for wasmi 0.9
  - Drives parser manually to get refined costs rather than delegating to wasmi
  - Added calibration and expanded refined costs a bit more
  - Added checks for complexity of const exprs
  - Protocol gated
  - feature="next" gated

This is getting close to where I want to merge it, I think it's a fairly good set of perf wins.

dmkozh commented on soroban-env-host/src/vm/parsed_module.rs:
```diff
@@ -0,0 +1,550 @@
+                }
+            }
+        }
+        if data > costs.n_memory_pages.saturating_mul(0x10000) {
```
Where does 0x10000 come from? Maybe use a named constant for clarity?

graydon replied:
Fixed.

dmkozh commented on soroban-env-host/src/vm/parsed_module.rs:
```diff
@@ -0,0 +1,550 @@
+                            ScErrorType::WasmVm,
+                            ScErrorCode::InvalidInput,
+                            "Too many arguments or results in wasm export",
+                            &[],
```
Can we include the actual values into the error message? Also applies to most of the errors here (where possible)

graydon replied:
Fixed (a fair amount -- always room for improvement)

dmkozh reviewed (commented):
I've looked at the host code (not benchmarks) and it looks good to me, only left a few small comments.

I think we'll need to be able to re-observe the existing tests with "next" feature enabled and with protocol 21 enabled, but that doesn't need to belong to this PR.

jayz22 commented on soroban-env-host/src/vm/parsed_module.rs:
```diff
@@ -0,0 +1,550 @@
+        })
+    }
+
+    /// Parse the wasm blob into a [Module] and its protocol number, checking check its interface version
```
Nit: remove "check"

graydon replied:
Fixed.

sisuresh commented on soroban-env-host/src/vm.rs:
```diff
@@ -72,160 +97,41 @@ impl std::hash::Hash for Vm {
-            ContractCostType::VmInstantiation,
-            Some(module_wasm_code.len() as u64),
-        )?;
+        let _span = tracy_span!("Vm::instantiate");
```
It's hard to tell if the changes in here can't cause a divergence with the current protocol. It looks like a bunch of code was moved around and the ordering of possible errors has changed? Is the idea here that it doesn't matter where the error checks happen as long as they get done at some point because they aren't observable in the ledger?

graydon replied:
Yeah, the code motion is unfortunate. A few things have changed:

  - Parsing and instantiating are split up
  - The parsing code is moved to `parsed_module.rs`
  - Linking is only linking the necessary imports

As far as I know, all the same checks are happening, but those that were parse-related have moved with the parsing-related code and therefore _can_ happen earlier (in the new protocol). When replaying the old protocol I believe they _should_ all still happen at the same time and in the same order. I'll be testing this as much as I can in the next couple weeks.

graydon replied:
Yeah. All the test observations will also catch divergence.

sisuresh commented on soroban-env-host/src/vm/module_cache.rs:
```diff
@@ -0,0 +1,97 @@
+            ));
+        }
+        let parsed_module = Rc::new(ParsedModule::new(host, &self.engine, &wasm, cost_inputs)?);
+        self.modules.insert(contract_id.clone(), parsed_module);
```
Should this be a `metered_clone`? I see that contract id clones are metered in other parts of the code.

graydon replied:
We could. It's a constant-sized term that's a mandatory part of work being done here which is never going to be zero-cost itself, so it probably doesn't matter much. But I might as well. This also reminded me to switch the BTreeMap to MeteredOrdMap.

graydon replied:
Done.

graydon commented on soroban-env-host/src/host/data_helper.rs:
```diff
@@ -122,15 +123,25 @@ impl Host {
+                let code = e.code.metered_clone(self)?;
+                let costs = match &e.ext {
+                    ContractCodeEntryExt::V0 => None,
+                    ContractCodeEntryExt::V1(v1) => Some(v1.cost_inputs.clone()),
```
No, it's a fixed-size object, incurs constant overhead.

jayz22 replied:
Yeah but we would like to be tight where possible. 
This struct is in the range of hundreds of bytes and may grow in the future. I think it's just safer to impl metered clone for it now. Most of the existing `DeclaredSize` types are in similar or lower size ranges and used in constant or one-off contexts.

graydon replied:
The struct is about 40 bytes (depending on how we meter it -- there's some complexity around ExtensionPoints we should discuss) but ok, I'll give it a try.

graydon replied:
Done, with horrible stuff to deal with ExtensionPoints

graydon commented on soroban-env-host/benches/common/cost_types/wasm_insn_exec.rs:
```diff
@@ -39,6 +42,138 @@ pub fn wasm_module_with_4n_insns(n: usize) -> Vec<u8> {
+    let mut me = ModEmitter::default();
+    for i in 0..n {
+        me.import_func("t", format!("_{i}").as_str(), Arity(0));
+    }
```
Fixed in subsequent changes.

graydon commented on soroban-env-host/benches/common/cost_types/vm_ops.rs:
```diff
@@ -1,45 +1,150 @@
+                }
+            }
+        }
+    };
```
Fixed (_random is actually the one you have to implement, but _best and _worst are redundant).

graydon commented on soroban-env-host/src/host/data_helper.rs:
```diff
@@ -122,15 +123,32 @@ impl Host {
+                match &e.ext {
+                    crate::xdr::ContractCodeEntryExt::V0 => (),
+                    crate::xdr::ContractCodeEntryExt::V1(v1) => {
+                        costs = VersionedContractCodeCostInputs::V1(v1.cost_inputs.clone())
```
No, it's fixed-size / constant overhead. We have lots of unmetered constant overheads.

graydon commented on soroban-env-host/src/vm/parsed_module.rs:
```diff
@@ -0,0 +1,550 @@
+                        return Err(host.err(
+                            ScErrorType::WasmVm,
+                            ScErrorCode::InvalidInput,
+                            "Too many arguments or results in wasm export",
```
...ok, this is not consistently done all through the codebase and I kinda don't want to include a change to a ton of unrelated files for it right now, but I'll fix instances in this PR?

dmkozh replied:
Sure, let's keep this PR focused - I'm only concerned here about the new code.

graydon commented on soroban-env-host/src/budget/model.rs:
```diff
@@ -81,7 +81,7 @@ impl From<f64> for ScaledU64 {
 }
 
-#[derive(Clone, Debug, Default)]
+#[derive(Clone, Copy, Debug, Default)]
```
In general I do not want structs to be Copy. In this specific case it's not just a matter of having to call `clone()` somewhere, it's needed to allow the nontrivial array initializer, and the only workarounds for array-initializing a non-Copy array are all quite convoluted and awful.

graydon replied:
(See the change in BudgetDimension::Default that now manually initializes the `cost_models` field, because it's now longer than 32 elements and the `Derive(Default)` only works up to 32 elements. Silly causal chain here.)

graydon commented on soroban-env-host/src/budget.rs:
```diff
@@ -23,7 +23,7 @@ use crate::{
 use dimension::{BudgetDimension, IsCpu, IsShadowMode};
 
-#[derive(Debug, Clone, Default, PartialEq, Eq, PartialOrd, Ord)]
+#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord)]
```
Same issue as above: the initializer for the BudgetTracker::cost_tracker field now only works if the CostTracker is Copy.

graydon commented:
Updated to address all comments (and all revisions from CAP review process today)

graydon commented on soroban-env-host/src/vm/parsed_module.rs:
```diff
@@ -0,0 +1,568 @@
+                                &[],
+                            ));
+                        }
+                        if mem.initial > 0xffff {
```
Oh, yeah, er .. I think this is a typo on my part and actually it should (a) be 0x10000 and (b) can be derived from the page size. It's just that it's a 32 bit address space and the pages are 64k. So 64k of them are allowed to fit. But not 64k-1 of them.

dmkozh replied:
Should this be replaced by `u32::MAX / WASM_STD_MEM_PAGE_SIZE_IN_BYTES` then?

graydon replied:
Something like that. I've fixed it in the updated code.

jayz22 reviewed (approved):
Just pointed out a few minor things that are can be addressed later.

jayz22 commented on soroban-env-host/src/vm/parsed_module.rs:
```diff
@@ -0,0 +1,577 @@
+                host.charge_budget(ContractCostType::VmInstantiation, Some(*wasm_bytes as u64))?;
+            }
+            #[cfg(feature = "next")]
+            Self::V1(inputs) => {
```
Added a tracking issue here https://github.com/stellar/rs-soroban-env/issues/1370

graydon replied:
commented in the bug -- don't think this one is worth the effort

jayz22 commented on soroban-env-host/src/budget.rs:
```diff
@@ -368,6 +409,107 @@ impl Default for BudgetImpl {
+                }
+                #[cfg(feature = "next")]
+                ContractCostType::InstantiateWasmDataSegmentBytes => {
+                    cpu.const_term = 25191;
```
Tracking issue https://github.com/stellar/rs-soroban-env/issues/1371

graydon replied:
fixed

jayz22 commented on soroban-env-host/src/vm/module_cache.rs:
```diff
@@ -0,0 +1,100 @@
+                &[],
+            ));
+        }
+        let parsed_module = Rc::new(ParsedModule::new(host, &self.engine, &wasm, cost_inputs)?);
```
Tracking issue https://github.com/stellar/rs-soroban-env/issues/1372

graydon replied:
fixed

jayz22 commented:
Added https://github.com/stellar/rs-soroban-env/issues/1374 to follow up on setting up `ModuleCache`, calibrating its cost, and charging for it in the hit path.

graydon commented on soroban-env-host/src/budget.rs:
```diff
@@ -368,6 +409,107 @@ impl Default for BudgetImpl {
+                }
+                #[cfg(feature = "next")]
+                ContractCostType::ParseWasmDataSegmentBytes => {
+                    cpu.const_term = 66075;
```
fixed

graydon commented on soroban-env-host/src/budget.rs:
```diff
@@ -471,6 +613,108 @@ impl Default for BudgetImpl {
+                }
+                #[cfg(feature = "next")]
+                ContractCostType::ParseWasmDataSegmentBytes => {
+                    mem.const_term = 17580;
```
fixed

graydon commented on soroban-env-host/src/budget.rs:
```diff
@@ -471,6 +613,108 @@ impl Default for BudgetImpl {
+                }
+                #[cfg(feature = "next")]
+                ContractCostType::InstantiateWasmDataSegmentBytes => {
+                    mem.const_term = 69256;
```
fixed
